### DISCUSSION TITLE
tutorials/export/exporting_for_android

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#15] [User-contributed notes]
# tutorials/export/exporting_for_android

Exporting for Android has fewer requirements than compiling Godot for Android. The following steps detail what is needed to set up the Android SDK and the engine. Install OpenJDK 17: Download and i...

https://docs.godotengine.org/en/latest/tutorials/export/exporting_for_android.html

**Comment by Swarkin** [comment]
1. When using the the sdkmanager command line tool to install everything, just specify that root folder as the Android SDK and it'll work.
2. Make sure to accept the prompt on the device to allow USB debugging.
3. If you get the error "All 'apksigner' tools located in Android SDK [...] failed to execute.", try running Godot as admin. When installing Java, there is an option to automatically set the JAVA_HOME variable. Make sure to check that as well. (See comment by @thinkpace)
4. To see output logs, simply enable `Deploy with Remote Debug` in the `Debug` panel at the top.

**Reply by thinkpace** [reply]
I experienced issue 3 "All 'apksigner' tools located in Android SDK [...] failed to execute." on my machine, adding JAVA_HOME as environment variable fixed my issue.

**Reply by ascyrax** [reply]
I am getting the same issue. 
where did you add the JAVA_HOME variable? in the terminal configuration file (.zshrc) or the godot editor ??
@thinkpace

**Reply by thinkpace** [reply]
@ascyrax my issue was with Windows (can be set in system settings), with Linux/Mac OS I assume .zshrc should work. You need to ensure that variable is loaded ("echo $JAVA_HOME" must print Java path). I'm not sure about Godot behaviour, if you're lucky it just uses the environment variable without the need to change anything in Godot config (as it did with Windows). 🙂

I suggest running Godot from command line after ensuring JAVA_HOME is set, this way you can be sure Godot has access to loaded environment variable.

**Reply by ascyrax** [reply]
yea, the variable is loaded. Its just that godot is somehow not able to access it.
i will try running it from command line. It could work :)
@thinkpace thank you :)

**Reply by ascyrax** [reply]
@thinkpace running the godot app from the command line solved the issue. thanks * 3000 :)

**Comment by wargaming0009** [comment]
I have this error even I have correct user and password. Please note that this is working in previous version.
editor/export/editor_export_platform.h:182 - Export: Release Username and/or Password is invalid for the given Release Keystore
0 param: --xr_mode_regular
1 param: --use_immersive

**Comment by RozyZx** [comment]
Is there a reason why there's no command to generate debug keystore here? I thought I was it before in this page. Is it the same as the release keystore command?

**Reply by Calinou** [reply]
Debug keystores are now automatically generated by the editor since Godot 4.3. You don't need to do anything manually now :slightly_smiling_face:

**Comment by GameDevJustin** [comment]
I am a little confused because my project shows the debug keystore user and pass filled out in editor settings but not debug keystore itself and I have no idea how to set it up

**Reply by GameDevJustin** [reply]
nevermind I was able to figure it out

**Reply by Calinou** [reply]
@GameDevJustin Can you post the solution you've found? This will be helpful for others stumbling upon the same issue :slightly_smiling_face:

**Comment by flochocinco** [comment]
With default settings as Android Exporter, unchecking "Signed" make the apk invalid for my phone (error at installation). If I check back "Signed" install is OK

**Comment by leider4604** [comment]
When I try to export to apk, I get this, how do I solve it?

>configure project:
WARNING: the option setting 'android.overridePathCheck=true' is experimental.
the current default is 'false'.

**Reply by Calinou** [reply]
This is a warning which you can safely ignore to my knowledge. It doesn't prevent exporting the project.

**Reply by leider4604** [reply]
It lets me export the project on Windows, but not for Android.

**Reply by najiro8v** [reply]
edit the code in the android folder

**Comment by SanzharIz8C** [comment]
"Keystore could not be found, export not possible"

Nothing was said about Keystore. Help please

**Reply by najiro8v** [reply]
move the keystore file to your project's folder, that happened to me

**Reply by SanzharIz8C** [reply]
Sorry, but I didn't understand you. What is a key vault anyway? debugging keystore? I tried to do something according to the guides on YouTube, but everything there seemed outdated, nothing was clear. They used the site "https://docs.godot.community/tutorials/export/exporting_for_android.html", where the command for it was indicated, which is not here. The purpose there doesn't work. Or am I just stupid for understand

**Reply by najiro8v** [reply]
The release keystore move it to your project's folder
And change the path in the export options, it's all you need to do
If You need more examples can You send me a text by discord NAJIGAMES#7999

**Reply by SuryaBoy** [reply]
Same problem, did you solve it ??

**Reply by SuryaBoy** [reply]
Here is the video like that solves the problem made by my self https://youtu.be/40fy7nUsWmk

**Comment by ivan-resetnikov** [comment]


**Reply by najiro8v** [reply]
Why change to CI/CD? If using GUI works ?
To sign the app with the gui don't forget to uncheck the 
"Export with Debug"

**Reply by Xevion** [reply]
Because CI/CD is superior and is a very legitimate choice. I don't know what Ivan's comment was originally, but I don't like the implication that anyone should relegate themselves to a GUI or that we should encourage more GUI-centric solutions. Godot is already struggling enough with having any decent CI/CD solutions as it is.

**Reply by najiro8v** [reply]
Well i don't use the CI/CD because in don't knew about that tool, if You know the solution, just posted it to help him, My post was a alternative solution and not an attack, if he use other different tool and not the GUI thats not My matter

**Reply by ivan-resetnikov** [reply]
I appreciate the attempt to help, truly. The phrasing was a little off, but things like this happen, I'm not an exception either.
I resolved the issue, though my editor settings were irreversibly erased.
The origin post was per my mistake, and not the engine's. The keystore was in a directory that was protected by my windows install.

**Comment by SecAlanraFlurry** [comment]
editor/export/editor_export_platform.h:179 - Code Signing: All 'apksigner' tools located in Android SDK 'build-tools' directory failed to execute. Please check that you have the correct version installed for your target sdk version. The resulting APK is unsigned.

whats wrong? pls help me

**Reply by najiro8v** [reply]
just delete android folder in the project and try to reinstall.

**Comment by N8King2020** [comment]
Got this error while trying to export to my android:
editor/export/editor_export_platform.h:182 - Run: Could not install to device: adb: failed to install C:/Program Files (x86)/Steam/steamapps/common/Godot Engine/editor_data/cache/tmpexport.1739948360.apk: Failure [INSTALL_PARSE_FAILED_NO_CERTIFICATES: Failed to collect certificates from /data/app/vmdl244546322.tmp/base.apk: Attempt to get length of null array]
  Performing Streamed Install

not sure what I'm doing wrong, would appreciate the help!

**Comment by ZarkonPro** [comment]
Hi.

I've get this error:

FAILURE: Build failed with an exception.
  
  * What went wrong:
  Could not open cp_settings generic class cache for settings file '...\android\build\settings.gradle' (...\.gradle\caches\8.2\scripts\4q30sfjl9o9m0vidtnj8l2d8l).
  > BUG! exception in phase 'semantic analysis' in source unit '_BuildScript_' Unsupported class file major version 67


Whats wrong with my configuration?

Thanks in advance.

**Reply by Lenivex** [reply]
I had the same problem.
I have no idea what it was, but it helped me somehow:
I added to the beginning of the file `{PROJECT_DIRECTORY}/android/build/gradle.properties`

this line (specified the path to the JDK 17 directory):
`org.gradle.java.home=D:/programs/java17`

somehow the error went away after that

**Reply by Aroulis8** [reply]
I think it worked because it was using a older version of the jdk but I am not sure

**Comment by nino-juan** [comment]
Leaving a comment for anyone trying to make an Android Launcher apk. 

To build an android launcher app, you can just activate the `Show as Launcher App` (requires `Use Gradle Build`).  You can find this setting on `project> export` .... then enable `advanced options` for the `Show as Launcher App` toggle to become visible under the `Package` category of the 'options` tabs.

Does anyone know if there is a list of available export options somewhere on the documentation?

**Reply by Calinou** [reply]
> Does anyone know if there is a list of available export options somewhere on the documentation?

You can find descriptions for all the Android export options here: https://docs.godotengine.org/en/latest/classes/class_editorexportplatformandroid.html

Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10822

**Comment by jakeklassen** [comment]
Are we able to target Android SDK 35? I am using libs requiring this compile target. So far I haven't figured out how to do this. I have SDK 35 installed via the SDK manager.

**Reply by cezbloch** [reply]
This discussion should answer your question:
https://www.reddit.com/r/godot/comments/1lpqlwi/update_to_android_target_sdk_35_in_godot_4/

**Comment by ITamas78** [comment]
Hello!
I have Godot 4.4.1 on Win10 and I have downloaded the necesssary tools (Android Studio, etc.) to make .APKs. I have set everything up (I hope, properly) and Godot exports .APK without any error message, but I can't run this .APK in Leapdroid emulator nor on cellphone, because I get a "There was a problem parsing the package" error message.
What is wrong and what should I do? I could make .APKs without errors with the old Godot 3.3. and Leapdroid can use these files.

**Reply by kesha2326** [reply]
Godot 4.4 stable. Have the same issue. Before enabling the “Use Gradle Build” option, I had built the project for Android several times without any issues. Everything worked fine. But after enabling that option and building the project again, the game stopped loading. Godot no longer recognizes global variables and doesn't launch scenes that are declared like @export var scene: PackedScene = preload("...")

**Reply by ITamas78** [reply]
Hello Kesha2326!
In my Godot version the "Use Gradle Build" option is unchecked. Should I turn it on?

**Reply by ITamas78** [reply]
And what mean this message in the output?
"Could not find version of build tools that matches Target SDK, using 30.0.3"
Although it is not signed as an error or warning.

**Reply by ITamas78** [reply]
Finally, problem solved! I had to install Android SDK build-tools v.34 for Android Studio (although v.35 was already installed) and the APK works fine on Android system greater than v.6. (Leapdroid supports only 4.4.4 but as Godot documentation wrote, Godot 4.x needs at least Android 6).

**Comment by ShadowMan11682** [comment]
I am getting the Error #1 after exporting the apk... I am not being able to find the solution even after exploring the export, editor settings for hours...

**Comment by s-adione** [comment]
When setting to Android SDK path the documentation says:
> %LOCALAPPDATA%\Android\Sdk\
Yet I had an error when going into the export menu. Instead I had to write the full path, here in my case:
> C:\Users\adrien\AppData\Local\Android\Sdk

**Reply by Calinou** [reply]
Godot's editor settings don't try to expand environment variables, which is why you need to write the full path instead.

This could probably be implemented, but there are nontrivial considerations about this (e.g. environment variables on macOS are not honored in certain conditions). Not to mention the syntax is also OS-specific in the first place (`%var%` on Windows, `$var` on Linux/macOS).

**Comment by TheMIU** [comment]
If you export an Android `.apk` file from Godot and encounter the error message:

> "You can't install the app on your device"

This is often due to architecture incompatibility. By default, Godot may export the APK with only ARM64-v8a support. Some devices, especially older or lower-end models, only support ARMv7, which causes the installation to fail.

#### Solution:

To ensure compatibility with a wider range of devices:

1. Open your project in Godot.
2. Go to **Project > Export > Android**.
3. In the **Architectures** section, make sure **ARMv7** is checked.
4. Re-export the APK.
5. Transfer the new APK to your device and install it.

After including ARMv7 support, the APK should install successfully on most Android devices.

**Comment by llama-nl** [comment]
Doc for the `APK Expansion` is missing.

**Comment by kokjoo89** [comment]
>WARNING: editor/export/editor_export_platform.h:240 - Code Signing: Could not find keystore, unable to export.
I'm new in godot. If export without debug it will show this issue. What am i doing wrong?

**Reply by Calinou** [reply]
You need to specify a release keystore in the Android export preset. The automatically-generated debug keystore is only used when exporting with debug enabled.

If you're just testing things out and don't plan to publish the APK on an app store, you could specify the debug keystore credentials (taken from the Editor Settings) in the release keystore fields. I do this when I'm doing performance profiling, which requires the use of release builds to reflect performance in production.

If you are publishing on an app store, you should get the release keystore file and credentials from e.g. the Google Play developer interface.

**Comment by lancemgiles** [comment]
Is there any information about device compatibility? Everything works fine on my phone, but one of my testers is unable to install it from Google Play because their device isn't compatible. I'm trying to figure out if there is anything I can do about this. The only thing I can think of is that maybe the default minSDK (in the export window) is too high but the tester's device is only two years old and at minimum would have Android 12 running

**Reply by lancemgiles** [reply]
I should have read through these comments - I was not supporting ARM7.

**Comment by antzGames** [comment]
Is there any info either official or not, on how to enable R8/PROGUARD for enabling obfuscated code and reduced app size on the Gradle builds?

**Comment by Locker72** [comment]
After upgrading to 4.5.rc1 from a project with multiple successful gradle Android exports from 4.4.1, the export would hang while downloading gradle-8.11.1-bin.zip.  This was after triple checking all of the SDK Platforms/Tools.  Downloading the gradle-8.11.1-bin.zip directly and unzipping into the correct folder led to other export failures.  In the end, simply doing a single non-grade export to APK allowed the next gradle export to AAB succeed.

**Reply by lancemgiles** [reply]
gradle exports seem to be working with 4.5rc2

**Comment by dac514** [comment]
How often are the "sdkmanager" instructions updated? 
For example, there's a build-tools;35.0.1 now _(as well as 36.0.0 and 36.1.0)_
We can see several newer versions than the ones prescribed using `cmdline-tools\latest\bin\sdkmanager --list`
Is using the exact specified versions in the CLI example the best practice - not the patch versions?

**Reply by Calinou** [reply]
The version listed is just a recommendation, but you can install a newer version and it should still work.
